source/postprocess/geoid.cc

```diff
@@ -0,0 +1,975 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <aspect/utilities.h>
+#include <aspect/postprocess/geoid.h>
+#include <aspect/geometry_model/spherical_shell.h>
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace internal
+    {
+      /* Definitions for HarmonicCoefficients */
+
+      template <int dim>
+      HarmonicCoefficients<dim>::HarmonicCoefficients(const unsigned int max_degree)
+      {
+        // Allocate vectors of the appropriate size for 2D and 3D.  For 2D we
+        // need cylindrical harmonics, for 3D we need spherical harmonics.
+
+        unsigned int k= (dim == 2 ?
+                         2 * (max_degree + 1 ) :             //cylindrical harmonics
+                         (max_degree+1)*(max_degree+2)/2 ) ; //spherical harmonics
+        sine_coefficients.resize(k);
+        cosine_coefficients.resize(k);
+      }
+
+
+      /* Definitions for the MultipoleExpansion base class */
+
+      // Constructor.  Just need the initialize the coefficients struct.
+      template <int dim>
+      MultipoleExpansion<dim>::MultipoleExpansion(const unsigned int max_degree)
+        :
+        max_degree(max_degree),
+        coefficients(max_degree)
+      {}
+
+      template <int dim>
+      void
+      MultipoleExpansion<dim>::add_quadrature_point( const Point<dim> &, const double, const double )
+      {
+        AssertThrow(false, ExcNotImplemented() );
```

Comment:
Instead of implementing a function that always throws an exception, you could implement it as a pure virtual function in the header file (adding '= 0' to the declaration of the function in the header file). Or do you want to allow derived classes to not implement this function?